framework/OutStreams/PlotInterfaces/OptPath.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2017 Battelle Energy Alliance, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Created on April 6, 2021
+
+@author: talbpaul
+"""
+from collections import defaultdict
+
+import matplotlib.pyplot as plt
+from matplotlib.lines import Line2D
+
+from .PlotInterface import PlotInterface
+from utils import InputData, InputTypes
+
+class OptPath(PlotInterface):
+  """
+    Plots the path that variables took during an optimization, including accepted and rejected runs.
+  """
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for class "cls".
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for specifying the input of cls.
+    """
+    spec = super().getInputSpecification()
+    spec.addSub(InputData.parameterInputFactory('source', contentType=InputTypes.StringType,
+        descr=r"""The name of the RAVEN DataObject from which the data should be taken for this plotter.
+              This should be the SolutionExport for a MultiRun with an Optimizer."""))
+    spec.addSub(InputData.parameterInputFactory('vars', contentType=InputTypes.StringListType,
+        descr=r"""Names of the variables from the DataObject whose optimization paths should be plotted."""))
+    return spec
+
+  def __init__(self):
+    """
+      Init of Base class
+      @ In, None
+      @ Out, None
+    """
+    super().__init__()
+    self.printTag = 'OptPath Plot'
+    self.source = None      # reference to DataObject source
+    self.sourceName = None  # name of DataObject source
+    self.vars = None        # variables to plot
+    self.markerMap = {'accepted': 'go',
```

Comment:
I'd add 'first': 'yo',
             'final': 'mo'
you might pick different colors if you like :)

Reply:
I thought about that; since two people did, it must be worth doing. I added it as you recommended.

Reply:
Great, thanks, Paul. By the way, I have tried it on GA and it is working fine, it certainly can have some extra features like a degrading alpha with the population number to separate populations from each other, but I think I can do that later. The only thing that I think will be really beneficial if you can add a 2D contour plot, this also works with all optimizers and is certainly needed to explore the obj function or even the fitness in my case. so another subplot that combines every two variables. If you can add that it will be awesome.

Reply:
I did think about this a bit, but I think this is a fairly complicated addition. The user is going to have to specify how they want to combine variables, and then we don't have an actual contour to plot since we have a rather sparse sampling of the space itself.

I recommend this be considered as a different plotter, or different PR extending this plotter. I don't have the program bandwidth to tackle that second type of plot right now.

Reply:
I meant to do all the combination 2 variables at a time. About the sparsity, in most problems, if it will continue till convergence, it is most likely it won't be that sparse, and if the user wants a denser plot it is intuitive to increase the number of iterations. But I understand the bandwidth issue. I have sent two tests one for GA and one for SA via email, please add those and check them, then remove from .tex that it only works for GD, then this is ready to be merged at least for this single plot. In the future, I wish if we can add the convergence mp4s, they were really the most descriptive.

Reply:
I completely agree, the animations are very interesting to see, and I would love to have the opportunity to automate them in some way.

Perhaps GA is different, but in GD we sample a very small portion of the overall space, making the contour plots very sparse for most problems, especially with single trajectory. For all the animations I've shown of the optimizer previously, a heavy grid sampling was performed a priori which allowed the background contour plot to be made. There may be some options there. At any rate, we can pick that up in discussion for future work.

I look forward to seeing your examples, and will add them in as tests, and modify the docs.

Reply:
Thanks, Paul. Let me know once you are done. Should I merge? or do you need two reviewers?

Reply:
I was also trying to make the plot for objVar ('ans') be a semiology but it was not straightforward as the self._objectiveVar is not accessible there. This is because GA and Simulated annealing sometimes end up having very slight fluctuations that are hard to see on the regular y-axis.

Reply:
I have a vision for this, where somehow we zoom in (or semilog) somehow as a function of convergence on the solution. As you noted, though, it would take additional information from the user, or perhaps we could do a density clustering on distance from the final result for each variable, or something. Maybe a local variance as a distance to the solution? All great possible additions for the future; this isn't written in final form obviously.

I don't think I need two reviewers, and no design-bases changes are made here so your review should be plenty. I will note the tests cannot currently run due to the HPC being down, however, so I imagine we will have to wait for those.

Reply:
Why don't you add an issue describing this vision as a feature request maybe one day we can do it. Let me know when the tests are ready so that I can merge them.

Reply:
done